### PR TITLE
[V2.0] HDF Update and TopicHash Introduction

### PR DESCRIPTION
**Pull request by DyrellC**
# Description of change
Addresses #260  and #256. Adds a MAC to the HDF for additional verification as well as adding the new `TopicHash` struct. `CursorStore` now maps branches by `TopicHash` instead of `Topic`, and HDF's now contain the `Topic` as `TopicHash` format instead of plaintext. 

## Type of change
- Enhancement (a non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested
- Previous examples continue to work 
- Tests all continue to pass
- fmt/clippy

### CODE REVIEW

**Comment by DyrellC** [comment]
Originally the plan was to include #259  in this PR as well, but what I have found is that there is no reasonable automated scenario that a user that would need to read the `Keyload` would be able to sequence to and see that `Keyload` without having known the `Topic` of the branch first, thus defeating the purpose of carrying the `Topic` through the `Keyload` @arnauorriols

**Comment by DyrellC** [comment]
As a consequence of the above stated, it brings into question the need to store cursors by `TopicHash` at all, as the reason for that was to accomodate behavior brought forward by #259.

File: streams/src/api/user.rs [hunk 1]
```diff
@@ -163,7 +166,11 @@ impl<T> User<T> {
     }
 
     pub fn topics(&self) -> impl Iterator<Item = &Topic> + ExactSizeIterator {
-        self.state.cursor_store.topics()
+        self.state.topics.iter()
+    }
+
+    pub(crate) fn topic_by_hash(&self, hash: &TopicHash) -> Option<Topic> {
```
Comment (kwek20) [inline streams/src/api/user.rs:172]: Looking at the code, this didnt need to be cloned really. a reference is enough

**Review by kwek20 — COMMENTED** [review]
Small change and then good to go!

File: streams/src/api/user.rs (hunk 1 above)
Comment (DyrellC) [reply streams/src/api/user.rs:172]: If we use a reference we need to clone each time anyways to set the latest link for the branch each time a message is handled

**Pull request merged**